### PR TITLE
Logging for Ray preprocessing

### LINKED ISSUE #117: Log metrics during preprocessing
Should log when using cache_dataset and when using on-the-fly. 

In cache_dataset:

* Ideally it should log to some kind of pbar. Should log number of docs processed, number of tokens created, number of shards finished

In on_the_fly:

* should probably log to wandb and occasionally to stdout. Wandb logs should include docs processed, tokens created, shards finished
* Can do pbar if we figure out interaction with the training pbar

### PR DESCRIPTION
**Pull request by dlwh**
I had lost the TQDM stuff in the ray migration, and I wanted to add something back. 

I also wanted to switch to Rich since I think it will compose a bit better ultimately with simultaneous training (and it's prettier).

Also adds wandb logging, also for simultaneous training support

Fixes #117

### CODE REVIEW

**Comment by dlwh** [comment]
@yifanmai rfal

File: src/levanter/data/shard_cache.py [hunk 1]
```diff
@@ -507,6 +619,32 @@ def __iter__(self):
...
+
+        self._metrics_monitors.append(monitor)
+        if self._monitor_thread is None:
+            self._monitor_thread = threading.Thread(target=self._monitor_metrics)
+            self._monitor_thread.start()
+
+    def _monitor_metrics(self):
+        while True:
+            try:
+                metrics = ray.get(self._broker.updated_metrics.remote())
```
Comment (Ivan-Zhou) [inline src/levanter/data/shard_cache.py:639]: Got an error message here:

```
Traceback (most recent call last):
  File "/home/ivanzhou/levanter/src/levanter/data/shard_cache.py", line 642, in _monitor_metrics
    metrics = ray.get(self._broker.updated_metrics.remote())
  File "/home/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/ray/actor.py", line 1189, in __getattr__
    raise AttributeError(
AttributeError: 'ActorHandle' object has no attribute 'updated_metrics'
Exception in thread Thread-8 (_monitor_metrics):
Traceback (most recent call last):
  File "/home/ivanzhou/miniconda3/envs/levanter/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/home/ivanzhou/miniconda3/envs/levanter/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ivanzhou/levanter/src/levanter/data/shard_cache.py", line 649, in _monitor_metrics
    raise e
  File "/home/ivanzhou/levanter/src/levanter/data/shard_cache.py", line 642, in _monitor_metrics
    metrics = ray.get(self._broker.updated_metrics.remote())
  File "/home/ivanzhou/miniconda3/envs/levanter/lib/python3.10/site-packages/ray/actor.py", line 1189, in __getattr__
    raise AttributeError(
AttributeError: 'ActorHandle' object has no attribute 'updated_metrics'
```

Comment (Ivan-Zhou) [reply src/levanter/data/shard_cache.py:639]: I couldn't find the definition of this function in our codebase or online. Is this a custom function?

Comment (dlwh) [reply src/levanter/data/shard_cache.py:686]: botched merge, PTAL

**Review by Ivan-Zhou — APPROVED** [review]
CleanShot 2023-05-17 at 19 45 25
Looks good!

**Pull request merged**